kingpin/actors/aws/elb.py

```diff
@@ -232,3 +221,110 @@ def _execute(self):
         utils.clear_repeating_log(repeating_log)
 
         raise gen.Return()
+
+
+class UseCert(ELBBaseActor):
+
+    """Find a server cert in IAM and use it for a specified ELB."""
+
+    all_options = {
+        'name': (str, REQUIRED, 'Name of the ELB'),
+        'port': (int, 443, 'Port associated with the cert'),
+        'region': (str, REQUIRED, 'AWS region name, like us-west-2'),
+        'cert_name': (str, REQUIRED, 'Unique IAM certificate name, or ARN'),
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Set up additional IAM connection."""
+        super(UseCert, self).__init__(*args, **kwargs)
+        self.iam_conn = boto.iam.connection.IAMConnection(
+            aws_access_key_id=aws_settings.AWS_ACCESS_KEY_ID,
+            aws_secret_access_key=aws_settings.AWS_SECRET_ACCESS_KEY)
+
+    @concurrent.run_on_executor
+    @utils.exception_logger
+    @retry(retry_on_exception=aws_settings.is_retriable_exception)
+    def _check_access(self, elb):
+        try:
+            # A blank ARN value should have code 'CertificateNotFound'
+            # We're only checking if credentials have sufficient access
+            elb.set_listener_SSL_certificate(self.option('port'), '')
+        except BotoServerError as e:
+            if e.error_code == 'AccessDenied':
+                raise exceptions.UnrecoverableActorFailure(e)
+
+    @concurrent.run_on_executor
+    @utils.exception_logger
+    @retry(retry_on_exception=aws_settings.is_retriable_exception)
+    def _get_cert_arn(self, name):
+        """Return a server_certificate ARN."""
+
+        self.log.debug('Searching for cert "%s"...' % name)
+        try:
+            cert = self.iam_conn.get_server_certificate(name)
+        except BotoServerError as e:
+            raise CertNotFound(
+                'Could not find cert %s. Reason: %s' % (name, e))
+
+        # Get the ARN of this cert
+        arn = cert['get_server_certificate_response'].get(
+            'get_server_certificate_result').get(
+            'server_certificate').get(
+            'server_certificate_metadata').get('arn')
+
+        return arn
+
+    @concurrent.run_on_executor
+    @utils.exception_logger
+    @retry(retry_on_exception=aws_settings.is_retriable_exception)
+    def _use_cert(self, elb, arn):
+        """Assign an ssl cert to a given ELB.
+
+        Args:
+            elb: boto elb object.
+            arn: ARN for server certificate to use.
+        """
+
+        self.log.info('Setting ELB "%s" to use cert arn: %s' % (elb, arn))
+        try:
+            elb.set_listener_SSL_certificate(self.option('port'), arn)
+        except BotoServerError as e:
+            raise exceptions.UnrecoverableActorFailure(
```

Comment:
This should be `RecoverableActorFailure` so that the user can decide whether they care if the action failed or not.